File: src/address.ts
import { init, g, unwrap } from './init';
import type { AddressInfo, Network } from './types';

/** Bitcoin address encoding, decoding, and creation. */
export const address = {
  /** Decode a Bitcoin address string and return detailed info.
   *  Calls Go: btcutil.DecodeAddress() from btcutil. */
  async decode(addr: string, network: Network = 'mainnet'): Promise<AddressInfo> {
    await init();
    return unwrap<AddressInfo>(g().address.decode(addr, network));
  },

  /** Create a P2PKH address from a 20-byte pubkey hash (hex).
   *  Calls Go: btcutil.NewAddressPubKeyHash() from btcutil. */
  async fromPubKeyHash(hexHash: string, network: Network = 'mainnet'): Promise<string> {
    await init();
    return unwrap<string>(g().address.fromPubKeyHash(hexHash, network));
  },

  /** Create a P2SH address from a 20-byte script hash (hex).
   *  Calls Go: btcutil.NewAddressScriptHashFromHash() from btcutil. */
  async fromScriptHash(hexHash: string, network: Network = 'mainnet'): Promise<string> {
    await init();
    return unwrap<string>(g().address.fromScriptHash(hexHash, network));
  },

  /** Create a P2SH address by hashing a serialized script (hex).
   *  Calls Go: btcutil.NewAddressScriptHash() from btcutil. */
  async fromScript(hexScript: string, network: Network = 'mainnet'): Promise<string> {
    await init();
    return unwrap<string>(g().address.fromScript(hexScript, network));
  },

  /** Create a P2WPKH address from a 20-byte witness program (hex).
   *  Calls Go: btcutil.NewAddressWitnessPubKeyHash() from btcutil. */
  async fromWitnessPubKeyHash(
    hexProgram: string,
    network: Network = 'mainnet',
  ): Promise<string> {
    await init();
    return unwrap<string>(
      g().address.fromWitnessPubKeyHash(hexProgram, network),
    );
  },

  /** Create a P2WSH address from a 32-byte witness program (hex).
   *  Calls Go: btcutil.NewAddressWitnessScriptHash() from btcutil. */
  async fromWitnessScriptHash(
    hexProgram: string,
    network: Network = 'mainnet',
  ): Promise<string> {
    await init();
    return unwrap<string>(
      g().address.fromWitnessScriptHash(hexProgram, network),
    );
  },

  /** Create a P2TR (Taproot) address from a 32-byte witness program (hex).
   *  Calls Go: btcutil.NewAddressTaproot() from btcutil. */
  async fromTaproot(hexProgram: string, network: Network = 'mainnet'): Promise<string> {
    await init();
    return unwrap<string>(g().address.fromTaproot(hexProgram, network));
  },

  /** Create a P2PK address from a serialized public key (hex).
   *  Calls Go: btcutil.NewAddressPubKey() from btcutil. */
  async fromPubKey(hexPubKey: string, network: Network = 'mainnet'): Promise<string> {
    await init();
    return unwrap<string>(g().address.fromPubKey(hexPubKey, network));
  },
};

File: wasm-wrapper/address.go
//go:build js && wasm

package main

import (
	"syscall/js"

	"github.com/btcsuite/btcd/address/v2"
)

func addressDecode(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "address"); e != nil {
		return e
	}
	network := optString(args, 1, "mainnet")
	params, e := getNetwork(network)
	if e != nil {
		return e
	}

	addr, err := address.DecodeAddress(args[0].String(), params)
	if err != nil {
		return errfResult("decode address: %s", err)
	}

	info := map[string]any{
		"address":       addr.EncodeAddress(),
		"scriptAddress": bytesToJS(addr.ScriptAddress()),
		"isForNet":      addr.IsForNet(params),
	}

	switch a := addr.(type) {
	case *address.AddressPubKeyHash:
		info["type"] = "p2pkh"
		info["hash160"] = bytesToJS(a.Hash160()[:])
	case *address.AddressScriptHash:
		info["type"] = "p2sh"
		info["hash160"] = bytesToJS(a.Hash160()[:])
	case *address.AddressTaproot:
		info["type"] = "p2tr"
		info["witnessVersion"] = int(a.WitnessVersion())
		info["witnessProgram"] = bytesToJS(a.WitnessProgram())
	case *address.AddressWitnessPubKeyHash:
		info["type"] = "p2wpkh"
		info["witnessVersion"] = int(a.WitnessVersion())
		info["witnessProgram"] = bytesToJS(a.WitnessProgram())
	case *address.AddressWitnessScriptHash:
		info["type"] = "p2wsh"
		info["witnessVersion"] = int(a.WitnessVersion())
		info["witnessProgram"] = bytesToJS(a.WitnessProgram())
	case *address.AddressPubKey:
		info["type"] = "p2pk"
		info["pubKeyFormat"] = int(a.Format())
	default:
		info["type"] = "unknown"
	}

	return okResult(info)
}

func addressFromPubKeyHash(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexHash"); e != nil {
		return e
	}
	hash, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	params, e := getNetwork(optString(args, 1, "mainnet"))
	if e != nil {
		return e
	}
	addr, err := address.NewAddressPubKeyHash(hash, params)
	if err != nil {
		return errfResult("fromPubKeyHash: %s", err)
	}
	return okResult(addr.EncodeAddress())
}

func addressFromScriptHash(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexHash"); e != nil {
		return e
	}
	hash, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	params, e := getNetwork(optString(args, 1, "mainnet"))
	if e != nil {
		return e
	}
	addr, err := address.NewAddressScriptHashFromHash(hash, params)
	if err != nil {
		return errfResult("fromScriptHash: %s", err)
	}
	return okResult(addr.EncodeAddress())
}

func addressFromScript(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexScript"); e != nil {
		return e
	}
	script, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	params, e := getNetwork(optString(args, 1, "mainnet"))
	if e != nil {
		return e
	}
	addr, err := address.NewAddressScriptHash(script, params)
	if err != nil {
		return errfResult("fromScript: %s", err)
	}
	return okResult(addr.EncodeAddress())
}

func addressFromWitnessPubKeyHash(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexProgram"); e != nil {
		return e
	}
	prog, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	params, e := getNetwork(optString(args, 1, "mainnet"))
	if e != nil {
		return e
	}
	addr, err := address.NewAddressWitnessPubKeyHash(prog, params)
	if err != nil {
		return errfResult("fromWitnessPubKeyHash: %s", err)
	}
	return okResult(addr.EncodeAddress())
}

func addressFromWitnessScriptHash(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexProgram"); e != nil {
		return e
	}
	prog, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	params, e := getNetwork(optString(args, 1, "mainnet"))
	if e != nil {
		return e
	}
	addr, err := address.NewAddressWitnessScriptHash(prog, params)
	if err != nil {
		return errfResult("fromWitnessScriptHash: %s", err)
	}
	return okResult(addr.EncodeAddress())
}

func addressFromTaproot(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexProgram"); e != nil {
		return e
	}
	prog, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	params, e := getNetwork(optString(args, 1, "mainnet"))
	if e != nil {
		return e
	}
	addr, err := address.NewAddressTaproot(prog, params)
	if err != nil {
		return errfResult("fromTaproot: %s", err)
	}
	return okResult(addr.EncodeAddress())
}

func addressFromPubKey(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexPubKey"); e != nil {
		return e
	}
	pub, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	params, e := getNetwork(optString(args, 1, "mainnet"))
	if e != nil {
		return e
	}
	addr, err := address.NewAddressPubKey(pub, params)
	if err != nil {
		return errfResult("fromPubKey: %s", err)
	}
	return okResult(addr.EncodeAddress())
}
